### PR TITLE
patrons: use marshmallow for API control

### PR DESCRIPTION
**Pull request by zannkukai**
Use Marshmallow schema to control `Patron` resources through API. The
`role` field is now controlled to limit changes depending to current
connected user.

Co-Authored-by: Renaud Michotte <renaud.michotte@gmail.com>

### CODE REVIEW

File: rero_ils/modules/patrons/permissions.py [hunk 1]
```diff
@@ -67,6 +70,27 @@ def needs(self, record=None, *args, **kwargs):
...
+
+    def excludes(self, record=None, **kwargs):
+        """Disallow operation check.
+
+        :param record; the record to check.
+        :param kwargs: extra named arguments.
+        :returns: a list of Needs to disable access.
+        """
+        roles = set(record.get('roles', []))
+        if not validate_role_changes(current_user, roles, raise_exc=False):
```
Comment (rerowep) [inline rero_ils/modules/patrons/permissions.py:89]: Only as a suggestion for one line:
```python
return [] if validate_role_changes(current_user, roles, raise_exc=False) else [any_user]
```

**Review by rerowep — APPROVED** [review]

File: rero_ils/modules/patrons/schemas/json.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        db_record = Patron.get_record_by_pid(data.get('pid'))
+        if not db_record:
+            abort(404)
+
+        # Check if `roles` of the patron changed. If not, we can stop
+        # the validation process.
+        original_roles = set(db_record.get('roles', []))
+        data_roles = set(data.get('roles', []))
+        role_changes = original_roles.symmetric_difference(data_roles)
+        if not role_changes:
```
Comment (jma) [inline rero_ils/modules/patrons/schemas/json.py:132]: Detecting if the roles have been changed seems as complex as `validate_role_changes`. Is it really required?

File: rero_ils/modules/patrons/utils.py [hunk 3]
```diff
@@ -40,15 +40,31 @@ def get_patron_pid_by_email(email):
...
+    config_roles = current_app.config.get(key_config, {})
+    manageable_roles = set()
+    for role in user.roles:
+        manageable_roles = manageable_roles.union(
+            config_roles.get(role.name, {}))
+    # If any difference are found between both sets, disallow the operation
+    if role_diffs := changes.difference(manageable_roles):
+        if raise_exc:
+            error_roles = ', '.join(role_diffs)
+            raise ValidationError(f'Unable to manage role(s) : {error_roles}')
```
Comment (jma) [inline rero_ils/modules/patrons/utils.py:66]: no space before a `:` in english I guess.

File: rero_ils/modules/users/api.py [hunk 4]
```diff
@@ -161,7 +161,7 @@ def _validate(cls, data, **kwargs):
         return data
 
     @classmethod
-    def get_by_id(cls, user_id):
+    def get_record(cls, user_id):
         """Get a user by a user_id.
```
Comment (jma) [inline rero_ils/modules/users/api.py:165]: You can remove this as it is inherited form invenio.

**Review by jma — APPROVED** [review]

**Pull request merged**